src/NuGet.Core/NuGet.Protocol/Converters/JsonExtensions.cs

```diff
@@ -14,6 +14,8 @@ public static class JsonExtensions
     {
         public const int JsonSerializationMaxDepth = 512;
 
+        public static readonly MetadataReferenceCache MetaCache = new MetadataReferenceCache();
```

Comment:
I'm not sure that a static cache is the best approach here. 

The challenge here is that this cache is only cleaned up when the solution is unloaded. If I keep searching for different package id and package versions then the cache will keep expanding. 
If I close the PM UI, it'll still retain this data into memory and that's not ideal.

Reply:
What is the event for PM UI close? I can add logic there. Currently I call  [this ](https://github.com/NuGet/NuGet.Client/blob/3328c404204c714787d55d736833042f53212d3c/src/NuGet.Core/NuGet.Protocol/Utility/MetadataReferenceCache.cs#L110) method to reset dictionary when we close solution.

Reply:
I am not a fan of statics, especially public ones (and I am a hypocrite because if you search the commit history, I added one). Plus I question every public API in `src\NuGet.Core`.  But especially in this case, which is a cache and can grow over time. Clearing it when the customer closes PM UI doesn't feel like a good solution because the user might have multiple PM UI windows open for different projects in the solution. Also, customers may be using PMC, not PM UI. I don't know how many of `NuGet.Protocol`'s resources use the metadata cache, but if restore uses it, then if a customer restores a large solution, but never opens/closes PM UI, the cache will keep all the references and therefore "leak" memory.

I think it would be better to have a `MetadataReferenceCacheResource` using `NuGet.Protocol`'s resource-provider/dependency injection model. 

Assuming that Visual Studio creates the `SourceRepository` objects, it can then do something like:

```cs
var metadataReferenceCache = new MetadataReferenceCache();
foreach (var sourceRepository in allSourceRepositories)
{
    var metadataReferenceCacheResource = await sourceRepository.GetResourceAsync<MetadataReferenceCacheResource>();
    metadataReferenceCacheResource.SetInstance(metadataReferenceCache);
}
```

and then `NuGet.Protocol` resources like `PackageMetadataResourceV3` can do
```cs
var metadataReferenceCacheResource = await sourceRepository.GetResourceAsync<MetadataReferenceCacheResource>();
var cache = metadataReferenceCacheResource.GetInstance(metadataReferenceCache);
```

Look at how most of the V3 resource providers use the ServiceIndexResource.

Reply:
@zivkan  Sure, I can refactor the code. 
Currently question is is it worth to pursue this effort. Because gain is slim even on large solution, but adds unnecessary runtime costs on all solutions regardless of solution size. For small solutions there will be almost 0 deduplication on this code path, because most of duplications exist on dependency graph which this task doesn't cover.

Reply:
@nkolev92 @zivkan Please check PR description for my updated investigation details. 

![image](https://user-images.githubusercontent.com/8766776/87742098-e4753900-c79a-11ea-8c62-ba683611e9fc.png)

Reply:
I think that the data your provided suggests an implementation bug, rather than invalidating the hypothesis that this technique is beneficial.

In your "PM UI closed" table, before there were zero MetadataReferenceCache instances, but after there is 1. We know that the MetadataReferenceCache holds instances of NuGetVersion and VersionRange. Therefore, by not allowing the cache to be garbage collected, we're also not allowing everything it's caching to be garbage collected.

Secondly, thanks to the bug where MetadataReferenceCache is not GCed, we can see that it probably holds around 2,300 NuGetVersions and 300 VersionRanges.  Since the "PM UI open" table contains more than 10 times this number, it suggests there's a bug and the implementation is not correctly using the MetadataReferenceCache for all instances of NuGetVersion or VersionRange.

Reply:
@zivkan regardless of that implementation bug end result is same it's kind of noise, we're not gaining anything we wanted when `PM UI` is `open`.  I think it's not worth to continue this. Because after PM UI is `closed` is already handled by my other #[9757](https://github.com/NuGet/Home/issues/9757)). 
About bug it's just MetadataReferenceCache is still in memory as long as SourceRepository is alive, I'm not sure when SourceRepository get GCed. 
If you remember we tied this cache to [SourceRepository](https://github.com/NuGet/NuGet.Client/blob/295057a1e7616880104d1978a5cb1b885d8d7519/src/NuGet.Core/NuGet.Protocol/SourceRepository.cs#L25). But it's not helping when PM UI is `open`.
![image](https://user-images.githubusercontent.com/8766776/87810415-cf87bc80-c811-11ea-8a5a-638d35e15992.png)

Reply:
> regardless of that implementation bug end result is same it's kind of noise, we're not gaining anything we wanted when `PM UI` is `open`.

This is what I'm saying might be a bug. I now realise that I misunderstood the issue. I thought it was to the use the cache everywhere, whereas now I see that it's only using it for two resource types.

I don't know if you have a hypothesis on where the 23,000 instances of NuGetVersion are being created from, but if they're coming from search results, then in the PM UI open case, I would expect the same number as the PM UI closed case, when the cache object is not GCed.

If there are 2300 NuGetVersion objects when PM UI is closed, but the cache is not GCed, then I would expect the same number when PM UI is open.  Since PM UI has 23,000 objects, this means either:
1. we can get reduce the number of NuGetVersion instances by 90% when PM UI is open
2. We should have 10x as many NuGetVersion instances when PM UI is closed and the cache is not GCed
3. the "real" value is somewhere in between.

However, if 20,000 instances of NuGetVersion are being created somewhere other than search results, then it explains why the cache was not helpful.

> About bug it's just MetadataReferenceCache is still in memory as long as SourceRepository is alive, I'm not sure when SourceRepository get GCed.

Having a known bug in your implementation means your results are not valid and therefore you cannot draw conclusions from the data.

Reply:
About bug it's just MetadataReferenceCache is still in memory as long as SourceRepository is alive, I'm not sure when SourceRepository get GCed.
> 
> Having a known bug in your implementation means your results are not valid and therefore you cannot draw conclusions from the data.

@zivkan  Sorry for not communicating well. As this task evolves ([#9757](https://github.com/NuGet/Home/issues/9757)) then actual problem was more of PM UI rather than duplicates. I didn't realize this first.

I created separate task for SourceRepository [memory leak](https://github.com/NuGet/Home/issues/9814). 
It looks like it's similar problem to  [#9757](https://github.com/NuGet/Home/issues/9757) where it can't release memory due to some implicit binding somewhere. Can I close this ticket and solve SourceRepository [memory leak](https://github.com/NuGet/Home/issues/9814) during next sprint?

Reply:
Yes, if this PR is not being actively worked on, I would be happy for it to be closed and reopened (or a new one) when it's picked up again.

Reply:
Thank you both for digging into this. 

* The PM UI Closed numbers clearly suggest that the reference cache is leaked and further investigatigation shows that the SourceRepository is the problem.
* The SourceRepository seemingly leaked is a concern, and I think we should fix that first, so definitely ok putting this on pause until then (consider moving it to draft). 
* The numbers in the PM UI open scenarios are weird. There's no theoretical reason why we have *more* VersionRange objects after implementing deduping than before deduping. To me this suggest there's either a bug in the implementation or just the soruce repository leak is skewing the numbers. Ofc a possibility could be different versions being available on the server, but I'm hoping you accounted for that :)

So @erdembayar I think you should do the following: 

1. Look into the SourceRepository leak, find the root cause and fix it. 
1. Once that's completed and merged, revisit this PR and get the same numbers with the PM UI open/closed. No need to numbers with PackageMetadataResource only. I think it's good enough to just see a "before duplication", "after duplication" change.

Reply:
> So @erdembayar I think you should do the following:
> 
> 1. Look into the SourceRepository leak, find the root cause and fix it.
> 2. Once that's completed and merged, revisit this PR and get the same numbers with the PM UI open/closed. No need to numbers with PackageMetadataResource only. I think it's good enough to just see a "before duplication", "after duplication" change.

@nkolev92 

1. I already separate task for it #9814 and I'll fix it during coming sprints. We already over half of this sprint and this fix is not planned for this sprint.
2. We already have that, please look "Before column", that is without any deduplication and other 2 are with deduplication.
Please note: That SourceRepository bug matter for PM UI is closed, but deduplication most matters when PM UI is `open`.
Without any deduplication PM UI `closed` is already good just 40KBs for this large solution.
 
![image](https://user-images.githubusercontent.com/8766776/87825233-965c4600-c82b-11ea-8c86-3dc68a8f77bf.png)

Reply:
Look at the version ranges at "before" and "after".

Reply:
Here comparison without any caching vs with caching only open PM UI Consolidate tab, not other tabs. 
Please note all memory dumping made after force GCed to get real picture of what is still in memory.
Even we deduplicated we have MetadataReferenceCache with 21258 bytes retained because it has dictionaries and dictionaries have overhead cost(pointer to actual item) for every item in it. We're bothered by this fine details because our gain was almost nothing and on edge, otherwise we don't need to bother such details.

![image](https://user-images.githubusercontent.com/8766776/87838986-b7ce2980-c84d-11ea-9c38-5dc798b7c14d.png)